### PR TITLE
Switch to event based/callback based approach for dream web

### PR DESCRIPTION
**Pull request by TesseractCat**
This uses HTTP fetch streaming to push images to the web interface as they are generated, and adds a progress bar by exposing the img_callback parameter to `prompt2image`.

As a result of using a callback based approach, when generating multiple images, it will add them to the result list immediately, rather than waiting until all images have completed.

### CODE REVIEW

File: scripts/dream_web.py [hunk 1]
```diff
@@ -56,44 +57,64 @@ def do_POST(self):
...
-        outputs = []
+        def image_done(image, seed):
+            config = post_data.copy() # Shallow copy
+            config['initimg'] = ''
+
+            # Write PNGs
+            pngwriter = PngWriter(
+                "./outputs/img-samples/", config['prompt'], 1
+            )
+            # metadata_str = f'prompt2png({json.dumps(config)} seed={seed}' # gets written into the PNG
```
Comment (tildebyte) [inline scripts/dream_web.py:68]: Is this being done elsewhere (not understanding why this is commented out)?

Comment (TesseractCat) [reply scripts/dream_web.py:68]: This variable is defined here: https://github.com/lstein/stable-diffusion/blob/main/ldm/simplet2i.py#L178, but never used. I'm not sure if it was planned for the future, but just for completeness sake I added it here as well.

**Comment by bakkot** [comment]
[Here](https://github.com/bakkot/stable-diffusion/commit/cbaa1c1445635939e688fed17546dc0f636e4e4b) is a commit on top of this branch which also streams the intermediate images, so you can watch it denoise. @TesseractCat feel free to pull this in to your branch.

This makes me want to have the ability to _cancel_ requests, because you can often see it's not going to be what you want before it actually finishes.

**Comment by lstein** [comment]
I'm fixing issues with the dream_web having to do with upscaling and will work on this next.

File: scripts/dream_web.py (hunk 1 above)
Comment (lstein) [reply scripts/dream_web.py:68]: This seems to be a hanging statement that should have been removed. Feel free to delete the comment. I will delete L178 from simplet2i.py.

**Review by lstein — COMMENTED** [review]
Unfortunately in the intervening time between submission of the PR, dream_web.py has been made into a module and moved into ldm/dream/server.py. scripts/dream.py --web now invokes this module. I think this makes better sense than maintaining two different launch scripts that share identical arguments.

Would it be possible to redo the PR using the new structure?

**Comment by bakkot** [comment]
I took the liberty of doing that rebase in https://github.com/lstein/stable-diffusion/pull/180. Feel free to ignore it if you'd prefer to keep this PR as the source of truth, though.

**Pull request closed without merging**